### ISSUE TITLE
Difference whether<v>exists in optional method

### ISSUE THREAD

**Issue by adam07018** [nus-cs2030/2122-s1#95]
Basically, when i first time write optional code, I write like this:

return Optional.of(circle);

However, when i was reviewing lecture, I found out prof write with extra <V> behind optional:

return Optional.<Circle>of(circle);

Both codes can run. What are the difference of two?

**Comment by adam07018** [comment]
> 
> 
> I don't see any difference. You mean `Optional.<Circle>of(circle);`?

Yes, what the hell does github automatically hide <>

**Comment by RussellDash332** [comment]
Might want to encase it inside markdown.

If it's not stated illegal then it's allowed, but of course if you want to play safe, you can actually not use it.
The hint is to look at the test cases which gives away the idea of using the new get method.

**Comment by FYimu** [comment]
> Btw, are we allow to use Optional.ofNullable() in lab? I am afraid that function name contains Null will fail in codecrunch

Allowed!

**Comment by Gibson0918** [comment]
I believe when Prof is doing `Optional.<Circle>of(circle)`. The `<Circle>` is a type witness. I found a [post](https://stackoverflow.com/questions/24932177/type-witness-in-java-generics) discussing it on stackOverflow.

**Comment by Enigmatrix** [comment]
Did he say type witness was recommended or must-use? (I've been using `Optional.ofNullable` without type witness lol)

**Comment by BrendanCheong** [comment]
> Did he say type witness was recommended or must-use? (I've been using `Optional.ofNullable` without type witness lol)

I've used type witness for Lab 6 with ```Optional.<V>ofNullable``` just to be consistent. I think its just best practice so that when you read the code you know exactly what is the type that is being passed in the Optional



> Btw, are we allow to use Optional.ofNullable() in lab? I am afraid that function name contains Null will fail in codecrunch

I think for lab 6 its necessary to use ```Optional.<V>ofNullable``` because you specifically want to allow null values in your Optional (turn null into Optional.empty) so that you can catch them and print the "No such Record" line

**Comment by Enigmatrix** [comment]
> I think for lab 6 its necessary to use Optional.<V>ofNullable because you specifically want to allow null values in your Optional (turns inputed null values into Optional.empty() so that you can catch them and print the "No such Record" line

Not really, we can use `map.containsKey(key)` on the `Map` to check if the key and value pair exists in the `Map`, and return `Optional.empty()` if so. Otherwise, return `Optional.of(map.get(key))`. This way we don't have to use `ofNullable`.

Thankfully, it seems the grader doesn't match `null` in a case insensitive manner, so all's good.
